Mathlib/Algebra/GroupPower/Ring.lean

```diff
@@ -0,0 +1,279 @@
+
+@[simp]
+theorem zero_pow_eq_zero [Nontrivial M] {n : ℕ} : (0 : M) ^ n = 0 ↔ 0 < n := by
+  constructor <;> intro h
+  · rw [pos_iff_ne_zero]
+    rintro rfl
+    simp at h
+  · exact zero_pow' n h.ne.symm
+
+theorem Ring.inverse_pow (r : M) : ∀ n : ℕ, Ring.inverse r ^ n = Ring.inverse (r ^ n)
+  | 0 => by rw [pow_zero, pow_zero, Ring.inverse_one]
+  | n + 1 => by
+    rw [pow_succ, pow_succ', Ring.mul_inverse_rev' ((Commute.refl r).pow_left n),
+      Ring.inverse_pow r n]
+
+end MonoidWithZero
+
+section CommMonoidWithZero
+
+variable [CommMonoidWithZero M] {n : ℕ} (hn : 0 < n)
+
+/-- We define `x ↦ x^n` (for positive `n : ℕ`) as a `monoid_with_zero_hom` -/
+def powMonoidWithZeroHom : M →*₀ M :=
+  { powMonoidHom n with map_zero' := zero_pow hn }
+
+@[simp]
+theorem coe_powMonoidWithZeroHom : (powMonoidWithZeroHom hn : M → M) = fun x ↦ (x^n : M) := rfl
+#align coe_pow_monoid_with_zero_hom coe_powMonoidWithZeroHom
+
+@[simp]
+theorem powMonoidWithZeroHom_apply (a : M) : powMonoidWithZeroHom hn a = a ^ n :=
+  rfl
+#align pow_monoid_with_zero_hom_apply powMonoidWithZeroHom_apply
+
+end CommMonoidWithZero
+
+theorem pow_dvd_pow_iff [CancelCommMonoidWithZero R] {x : R} {n m : ℕ} (h0 : x ≠ 0)
+    (h1 : ¬IsUnit x) : x ^ n ∣ x ^ m ↔ n ≤ m := by
+  constructor
+  · intro h
+    rw [← not_lt]
+    intro hmn
+    apply h1
+    have : x ^ m * x ∣ x ^ m * 1 := by
+      rw [← pow_succ', mul_one]
+      exact (pow_dvd_pow _ (Nat.succ_le_of_lt hmn)).trans h
+    rwa [mul_dvd_mul_iff_left, ← isUnit_iff_dvd_one] at this
+    apply pow_ne_zero m h0
+  · apply pow_dvd_pow
+
+section Semiring
+
+variable [Semiring R] [Semiring S]
+
+protected theorem RingHom.map_pow (f : R →+* S) (a) : ∀ n : ℕ, f (a ^ n) = f a ^ n :=
+  map_pow f a
+#align ring_hom.map_pow RingHom.map_pow
+
+theorem min_pow_dvd_add {n m : ℕ} {a b c : R} (ha : c ^ n ∣ a) (hb : c ^ m ∣ b) :
+    c ^ min n m ∣ a + b := by
+  replace ha := (pow_dvd_pow c (min_le_left n m)).trans ha
+  replace hb := (pow_dvd_pow c (min_le_right n m)).trans hb
+  exact dvd_add ha hb
+
+end Semiring
+
+section CommSemiring
+
+variable [CommSemiring R]
+
+theorem add_sq (a b : R) : (a + b) ^ 2 = a ^ 2 + 2 * a * b + b ^ 2 := by
+  simp only [sq, add_mul_self_eq]
+#align add_sq add_sq
+
+theorem add_sq' (a b : R) : (a + b) ^ 2 = a ^ 2 + b ^ 2 + 2 * a * b := by
+  rw [add_sq, add_assoc, add_comm _ (b ^ 2), add_assoc]
+#align add_sq' add_sq'
+
+alias add_sq ← add_pow_two
+
+end CommSemiring
+
+section HasDistribNeg
+
+variable [Monoid R] [HasDistribNeg R]
+
+variable (R)
+
+theorem neg_one_pow_eq_or : ∀ n : ℕ, (-1 : R) ^ n = 1 ∨ (-1 : R) ^ n = -1
+  | 0 => Or.inl (pow_zero _)
+  | n + 1 => (neg_one_pow_eq_or n).symm.imp
+    (fun h ↦ by rw [pow_succ, h, neg_one_mul, neg_neg])
+    (fun h ↦ by rw [pow_succ, h, mul_one])
+
+variable {R}
+
+theorem neg_pow (a : R) (n : ℕ) : (-a) ^ n = (-1) ^ n * a ^ n :=
+  neg_one_mul a ▸ (Commute.neg_one_left a).mul_pow n
+
+/- Porting note: bit0 and bit1 are deprecated
```

Comment:
We're still porting lemmas about bit0 and bit1, as we need some fraction of them later.